src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,330 @@
+    query: QueryDict, params: Dict[Any, Any]
+) -> Tuple[List[Tuple[str, str, List[Any]]], Dict[str, Any]]:
+    filter_keys = {}
+    conditions = []
+
+    for filter_name in DIMENSION_MAP:
+        raw_filter = query.getlist(filter_name, [])
+        resolved_filter = DIMENSION_MAP[filter_name].resolve_filter(raw_filter)
+        if len(resolved_filter) > 0:
+            conditions.append((filter_name, "IN", resolved_filter))
+
+    if "project_id" in params:
+        filter_keys["project_id"] = params["project_id"]
+    if "organization" in params:
+        filter_keys["organization_id"] = params["organization_id"]
+    return conditions, filter_keys
+
+
+class QueryDefinition:
+    """
+    This is the definition of the query the user wants to execute.
+    This is constructed out of the request params, and also contains a list of
+    `fields` and `groupby` definitions as [`ColumnDefinition`] objects.
+    """
+
+    def __init__(
+        self,
+        query: QueryDict,
+        params: Dict[Any, Any],
+        allow_minute_resolution: Optional[bool] = True,
+    ):
+        raw_fields = query.getlist("field", [])
+        raw_groupby = query.getlist("groupBy", [])
+        if len(raw_fields) == 0:
+            raise InvalidField('Request is missing a "field"')
+
+        self.fields = {}
+        self.aggregations = []
+        self.query: List[Any] = []  # not used but needed for compat with sessions logic
+        start, end, rollup = get_constrained_date_range(query, allow_minute_resolution)
+        self.dataset = _outcomes_dataset(rollup)
+        self.rollup = rollup
+        self.start = start
+        self.end = end
+
+        for key in raw_fields:
+            if key not in COLUMN_MAP:
+                raise InvalidField(f'Invalid field: "{key}"')
+            field = COLUMN_MAP[key]
+            self.aggregations.append(field.aggregation(self.dataset))
+            self.fields[key] = field
+
+        self.groupby = []
+        for key in raw_groupby:
+            if key not in GROUPBY_MAP:
+                raise InvalidField(f'Invalid groupBy: "{key}"')
+            self.groupby.append(GROUPBY_MAP[key])
+
+        if len(query.getlist("category", [])) == 0 and "category" not in raw_groupby:
+            raise InvalidQuery("Query must have category as groupby or filter")
+
+        query_columns = set()
+        for field in self.fields.values():
+            query_columns.update(field.get_snuba_columns(raw_groupby))
+        for groupby in self.groupby:
+            query_columns.update(groupby.get_snuba_columns())
+        self.query_columns = list(query_columns)
+
+        query_groupby = set()
+        for groupby in self.groupby:
+            query_groupby.update(groupby.get_snuba_groupby())
+        self.query_groupby = list(query_groupby)
+
+        self.conditions, self.filter_keys = get_filter(query, params)
+
+
+def run_outcomes_query(query: QueryDefinition) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
```

Comment:
Adding `ResultSet = List[Dict[str, Any]]` would clean up a lot of these functions.

Reply:
100%. updated! looks much better now.